backend/vectorize_utils.ml

```diff
@@ -95,3 +95,37 @@ module Vectorized_instruction = struct
       results = Array.init res_count (fun i -> Result i)
     }
 end
+
+let vectorizable_machtypes (r1 : Reg.t) (r2 : Reg.t) =
+  match r1.typ, r2.typ with
+  | (Vec128 | Valx2), (Val | Int | Addr | Float | Float32 | Vec128 | Valx2)
+  | (Val | Int | Addr | Float | Float32), (Vec128 | Valx2) ->
+    Misc.fatal_errorf "Unexpected vector machtype Vec128 or Valx2: %a %a"
+      Printreg.reg r1 Printreg.reg r2
+  | Val, Val -> true
+  | Val, (Int | Addr | Float | Float32) | (Int | Addr | Float | Float32), Val ->
+    false
+  | (Int | Addr | Float | Float32), (Int | Addr | Float | Float32) ->
+    (* It is safe to mix Float32, Float, and Int for the purpose of GC, because
+       they are not scannable. It may not be possible to vectorize the
+       operation. *)
+    true
+
+let vectorize_machtypes (pack : Reg.t list) : Cmm.machtype_component =
+  match pack with
+  | [] -> assert false
+  | hd :: tl -> (
+    let can_vectorize = List.for_all (vectorizable_machtypes hd) tl in
+    if not can_vectorize
+    then
+      Misc.fatal_errorf "register pack with incompatible mach types:"
+        Printreg.reglist pack;
+    match hd.typ, List.length pack with
+    | (Int | Addr | Float | Float32), _ ->
+      (* allows subregs, width should be correct by construction of [Group]. *)
```

Comment:
May as well check the length is 2 (or 4 for float32) since the val case does

Reply:
Done